x/mongo/driver/auth/oidc.go

```diff
@@ -0,0 +1,329 @@
+func (oa *OIDCAuthenticator) getAccessToken(
+	ctx context.Context,
+	args *OIDCArgs,
+	callback OIDCCallback,
+) (string, error) {
+	if oa.accessToken != "" {
+		return oa.accessToken, nil
+	}
+
+	cred, err := callback(ctx, args)
+	if err != nil {
+		return "", err
+	}
+
+	oa.accessToken = cred.AccessToken
+	oa.tokenGenID++
+	oa.cfg.Connection.SetOIDCTokenGenID(oa.tokenGenID)
+	if cred.RefreshToken != nil {
+		oa.refreshToken = cred.RefreshToken
+	}
+	return cred.AccessToken, nil
+}
+
+// TODO GODRIVER-3246: Implement OIDC human flow
+// This should only be called with the Mutex held.
+//func (oa *OIDCAuthenticator) getAccessTokenWithRefresh(
+//	ctx context.Context,
+//	callback OIDCCallback,
+//	refreshToken string,
+//) (string, error) {
+//
+//	cred, err := callback(ctx, &OIDCArgs{
+//		Version:      apiVersion,
+//		IDPInfo:      oa.idpInfo,
+//		RefreshToken: &refreshToken,
+//	})
+//	if err != nil {
+//		return "", err
+//	}
+//
+//	oa.accessToken = cred.AccessToken
+//	oa.tokenGenID++
+//	oa.cfg.Connection.SetOIDCTokenGenID(oa.tokenGenID)
+//	return cred.AccessToken, nil
+//}
+
+// invalidateAccessToken invalidates the access token, if the force flag is set to true (which is
+// only on a Reauth call) or if the tokenGenID of the connection is greater than or equal to the
+// tokenGenID of the OIDCAuthenticator. It should never actually be greater than, but only equal,
+// but this is a safety check, since extra invalidation is only a performance impact, not a
+// correctness impact.
+// This must only be called with the lock held
+func (oa *OIDCAuthenticator) invalidateAccessToken() {
+	tokenGenID := oa.cfg.Connection.OIDCTokenGenID()
+	if tokenGenID >= oa.tokenGenID {
+		oa.accessToken = ""
+		oa.cfg.Connection.SetOIDCTokenGenID(0)
+	}
+}
+
+// Reauth reauthenticates the connection when the server returns a 391 code. Reauth is part of the
+// driver.Authenticator interface.
+func (oa *OIDCAuthenticator) Reauth(ctx context.Context) error {
+	oa.mu.Lock()
+	oa.invalidateAccessToken()
+	oa.mu.Unlock()
+	// it should be impossible to get a Reauth when an Auth has never occurred,
+	// so we assume cfg was properly set. There is nothing to enforce this, however,
+	// other than the current driver code flow. If cfg is nil, Auth will return an error.
+	return oa.Auth(ctx, oa.cfg)
+}
+
+// Auth authenticates the connection.
+func (oa *OIDCAuthenticator) Auth(ctx context.Context, cfg *Config) error {
+	// the Mutex must be held during the entire Auth call so that multiple racing attempts
+	// to authenticate will not result in multiple callbacks. The losers on the Mutex will
+	// retrieve the access token from the Authenticator cache.
+	oa.mu.Lock()
+	defer oa.mu.Unlock()
+	var err error
+
+	if cfg == nil {
+		return newAuthError(fmt.Sprintf("config must be set for %q authentication", MongoDBOIDC), nil)
+	}
+	oa.cfg = cfg
+
+	if oa.accessToken != "" {
+		err = ConductSaslConversation(ctx, cfg, "$external", &oidcOneStep{
```

Comment:
Holding an exclusive lock while we run the SASL conversation will limit the auth part of connection establishment to one-at-a-time, overriding the intended behavior of `maxConnecting`. Instead, we should only hold a lock when invalidating and acquiring a new access token.

Unfortunately that means the strategy of storing the `auth.Config` on the `OIDCAuthenticator` also won't work.

E.g. changes to prevent locking during the SASL conversation and allow multiple `Config` instances:
```go
func (oa *OIDCAuthenticator) invalidateAccessToken(tokenGenID uint64) {
	oa.mu.Lock()
	defer oa.mu.Unlock()

	if oa.tokenGenID <= tokenGenID {
		oa.accessToken = ""
	}
	// ...
}

func (oa *OIDCAuthenticator) getAccessToken(...) (string, uint64, error) {
	oa.mu.Lock()
	defer oa.mu.Unlock()

	if oa.accessToken != "" {
		return oa.accessToken, oa.tokenGenID, nil
	}
	// ...
}

func (oa *OIDCAuthenticator) Auth(ctx context.Context, cfg *Config) error {
	var err error

	if cfg == nil {
		return newAuthError(fmt.Sprintf("config must be set for %q authentication", MongoDBOIDC), nil)
	}

	accessToken, tokenGenID, err := oa.getAccessToken(...)

	if oa.accessToken != "" {
		err = ConductSaslConversation(ctx, cfg, "$external", &oidcOneStep{
			userName:    oa.userName,
			accessToken: oa.accessToken,
		})
		if err == nil {
			return nil
		}
		oa.invalidateAccessToken(tokenGenID)
		time.Sleep(invalidateSleepTimeout)
	}

	oa.mu.Lock()
	defer oa.mu.Unlock()
	// ...
}
```

Reply:
I think this works, let me run the tests. My concern is machine_1_2, but I do think this works.

Reply:
This worked